packages/cosmic-swingset/src/fees.js

```diff
@@ -0,0 +1,133 @@
+import { assert, details as X } from '@agoric/assert';
+
+import { DecimalMath } from './rational.js';
+
+// This controls how precise the fee calculation is.
+export const DEFAULT_COUNT_DECIMAL_PLACES = 12n;
+
+// Logic adapted from @agoric/swingset-vat/src/vats/comms/delivery.js
+const countSlots = message => {
+  assert.typeof(message, 'string');
+  // seqNum:ackSeqNum:msgBody
+  const match = message.match(/^[^:]+:[^:]+:(.*)$/s);
+  assert(match);
+  const msgBody = match[1];
+  assert.typeof(msgBody, 'string');
+  if (msgBody.startsWith('deliver:')) {
+    // deliver:$target:$method:[$result][:$slots...];body
+    const sci = msgBody.indexOf(';');
+    assert(sci >= 0);
+    const [_command, _target, _method, _result, ...slots] = msgBody
+      .slice(0, sci)
+      .split(':');
+    return slots.length;
+  } else if (msgBody.startsWith('resolve:')) {
+    // resolve:fulfill;${target}${rmss};${resolution.body}
+    const subMessages = msgBody.split('\n');
+    let numSlots = 0;
+    subMessages.forEach(subMessage => {
+      const sci = subMessage.indexOf(';');
+      assert(sci >= 0);
+      const [_command, _rejectOrFulfill, _rpid, ...slots] = subMessage
+        .slice(0, sci)
+        .split(':');
+      numSlots += slots.length;
+    });
+    return numSlots;
+  } else if (msgBody.startsWith('gc:')) {
+    // gc:$type:$$rref
+    // TODO: Do we need anything other than message length?
+    return 0;
+  }
+
+  assert.fail(X`Unknown mailbox message type: ${msgBody}`);
+};
+
+export const calculateMailboxFee = (messages, params) => {
+  assert(Array.isArray(messages));
+  const feePerInboundTx = DecimalMath.fromString(params.feePerInboundTx);
+  const feePerMessage = DecimalMath.fromString(params.feePerMessage);
+  const feePerMessageSlot = DecimalMath.fromString(params.feePerMessageSlot);
+  const feePerMessageByte = DecimalMath.fromString(params.feePerMessageByte);
+
+  let fee = feePerInboundTx;
+
+  const messageCountFee = DecimalMath.multiply(messages.length, feePerMessage);
+  fee = DecimalMath.add(fee, messageCountFee);
+
+  // Break up the sequence and message.
+  messages.forEach(seqMessage => {
+    assert(Array.isArray(seqMessage));
+    const [_seq, message] = seqMessage;
+    assert.typeof(message, 'string');
+
+    // Calculate the slot cost.
+    const numSlots = countSlots(message);
+    const messageSlotCountFee = DecimalMath.multiply(
+      numSlots,
+      feePerMessageSlot,
+    );
+    fee = DecimalMath.add(fee, messageSlotCountFee);
+
+    const messageByteCountFee = DecimalMath.multiply(
+      message.length,
+      feePerMessageByte,
+    );
+    fee = DecimalMath.add(fee, messageByteCountFee);
+  });
+
+  return { fee: DecimalMath.toString(fee), denom: params.feeDenom };
+};
+
+export const makeUpdateChainAccountFee = (
```

Comment:
I used Nat for everything and removed the preemptive usage so that we can settle on a bean counter interface before SwingSet uses the `beans_per_unit` parameter.